VK editorial edits to lp chapter (#72)
drvinceknight/amwoss: src/chapters/08/main.tex

\end{figure}

Expression~\ref{eqn:objective_paint} corresponds to the total profit, which is
the value to be maximised. As a line in 2-dimensional space,
this expression fixes its gradient, but its value determines the size of the
\(y\)-intercept. Therefore optimising this function corresponds to pushing a line

> with that gradient to its furthest extreme within the feasible region,

Reviewer: The definition of `feasible region` was removed above, and it's interpretation as a convex region bounded by the constraints. Should we define that here?

Author: `The shaded region in the Figure ... shows the pairs of values of A and B which are feasible, that is: they satisfy the constraints.`

demonstrated in Figure~\ref{fig:paint_lp}.
Therefore for this problem the optimum occurs in a particular vertex of the
feasible region, at \(A = \frac{12}{7}\) and \(B = \frac{20}{7}\).

This works well as \(A\) and \(B\) can take any real value in the feasible region.
Some problems must be formulated as Integer Linear Programs where the decision
% TODO Check Programme versus Program for consistency (VK: I don't care which
% one)
variables are restricted to integers. There are a number of methods that can
help adapt a real solution to an integer solution. These include cutting
planes, which introduce new constraints around the real solution to force an
integer value; and branch and bound methods, where we iteratively convert
decision variables to their closest two integers and remove any infeasible
solutions.
% TODO Add footnote reference

Both Python and R have libraries that carry out the linear and integer
programming algorithms for us. When solving these kinds of problems, formulating
them as linear systems is the most important challenge.

Consider again the exam scheduling problem from Section~\ref{sec:problem} which

> will now be formulated as an Integer Linear Program.

Reviewer: ```suggestion
will now be formulated as an integer linear program.
```

respectively.
Therefore \(M = A_c \cup A_o \cup B_c \cup B_o \cup C_c \cup C_o\).

Additionally there are further clashes between these sets:
\begin{itemize}
  \item No modules in \(A_c \cup A_o\) can be scheduled together as they may
  share students, this is given by the constraint in inequality~\ref{eqn:clique1}.
  \item No modules in \(B_c \cup B_o \cup A_c\), can be scheduled together as
  they may share students, given by inequality~\ref{eqn:clique2}.
  \item No modules in \(B_c \cup B_o \cup C_o\), can be scheduled together as
  they may share students, given by inequality~\ref{eqn:clique3}.
  \item No modules in \(B_o \cup C_c \cup C_o\), can be scheduled together as
  they may share students, given by inequality~\ref{eqn:clique4}.
\end{itemize}

Define \(\{Y_t \text{ for } t \in T\}\) as a set of auxiliary binary
decision variables, where \(Y_t\) is 1 if time slot \(t\) is being used. This is
enforced by Inequality~\ref{eqn:auxiliary}.

Equation~\ref{eqn:schedule_all_modules},
ensures all modules are scheduled once and once only.
Thus altogether the integer program becomes:
% TODO Capitalize Integer Program or not? (VK I don't care, but need to check

> % for consistency)

Reviewer: I don't like capitalising things that don't need it.

Author: Fine by me, I think I only did it because you had capitalised some other stuff (but maybe that was me in the first place 😆 ).

Z_{|M|^2 + m} &= Y_m
\end{align}

where \(t\) and \(m\) are indices starting at 0.
For example \(Z_{17}\) would correspond to \(X_{3, 2}\), the decision variable
representing whether module number 4 is scheduled on day 3; \(Z_{208}\) would
correspond to \(Y_{12}\), the decision variable representing whether there is an
exam scheduled for day 12.

Parameters \(c\), \(A\), and \(b\) can be determined by using this same conversion
from the model in Equations~\ref{eqn:objective_modules} to \ref{eqn:clique4}.
The vector \(c\) would be \(|M|^2\) zeroes followed by \(|M|\) ones.
The vector \(b\) would be zeroes for all the rows representing
Equation~\ref{eqn:auxiliary}, and ones for all other constraints.

\section{Solving with Python}\label{sec:solving-with-python}

> In this book the Python library \mintinline{python}{pulp} will be used to

Reviewer: ```suggestion
In this book the Python library Pulp will be used to
```

Day 13: 8,
\end{pyout}

The order of the days do not matter here, but we 7 days are required in order to
schedule all exams with no clashes, with at most three exams scheduled each day.

\section{Solving with R}\label{sec:solving-with-R}

> The R package \mintinline{R}{ROI}, the R Optimization

Reviewer: ```suggestion
The R package ROI, the R Optimization
```

Infrastructure will be used here.
This is a library of code that acts as a front end to a number of other solvers
that need to be installed externally, allowing a range of optimisation problems
to be solved with a number of different solvers.
The solver that will be used here is called the CBC MILP Solver, which needs to
be installed as well as the R \mintinline{R}{rcbc} package.
% TODO Do we need to point at installation documentation for this?

> The \mintinline{R}{ROI} package requires that the linear program is

Reviewer: ```suggestion
The ROI package requires that the linear program is
Write
```

}
\end{Rin-no-test}

Finally the following function uses all previous functions
to assemble a coefficients matrix.
It loops though the parameters for each constraint row required, uses the
appropriate function to create the row of the coefficients matrix, sets the

> appropriate inequality direction (\(\leq$, $=$, $\geq\)), and the value of the

Reviewer: does this compile as expected?

Author: Yeah but not correctly, I automated a subsitution and broke it. I'll fix.

Reviewer: I'll fix when I go through and test the code :)

schedule <- get_schedule(
    sol = sol,
    n_days = n_days,
    n_modules = n_modules
    )

> cat(schedule)

Author: Interesting!
